File: ios/Runner/SuperGpsStreamOptions.swift
import Foundation

struct SuperGpsStreamOptions: Equatable {
    var updateIntervalMs: Int64 = 700
    var minUpdateIntervalMs: Int64 = 500
    var minUpdateDistanceMeters: Float = 0
    var enableBarometer: Bool = false

    static func from(arguments: Any?) -> SuperGpsStreamOptions {
        guard let map = arguments as? [String: Any] else {
            return SuperGpsStreamOptions()
        }

        return SuperGpsStreamOptions(
            updateIntervalMs: readInt64(map, key: "updateIntervalMs", default: 700),
            minUpdateIntervalMs: readInt64(map, key: "minUpdateIntervalMs", default: 500),
            minUpdateDistanceMeters: readFloat(map, key: "minUpdateDistanceMeters", default: 0),
            enableBarometer: readBool(map, key: "enableBarometer", default: false)
        )
    }

    private static func readBool(_ map: [String: Any], key: String, default defaultValue: Bool) -> Bool {
        guard let value = map[key] else { return defaultValue }
        return value as? Bool ?? defaultValue
    }

    private static func readInt64(_ map: [String: Any], key: String, default defaultValue: Int64) -> Int64 {
        guard let value = map[key] else { return defaultValue }
        if let intValue = value as? Int { return Int64(intValue) }
        if let intValue = value as? Int64 { return intValue }
        if let doubleValue = value as? Double { return Int64(doubleValue) }
        if let floatValue = value as? Float { return Int64(floatValue) }
        return defaultValue
    }

    private static func readFloat(_ map: [String: Any], key: String, default defaultValue: Float) -> Float {
        guard let value = map[key] else { return defaultValue }
        if let floatValue = value as? Float { return floatValue }
        if let doubleValue = value as? Double { return Float(doubleValue) }
        if let intValue = value as? Int { return Float(intValue) }
        if let intValue = value as? Int64 { return Float(intValue) }
        return defaultValue
    }
}

File: android/app/src/main/kotlin/com/sps/patrol/gps/SuperGpsStreamOptions.kt
package com.sps.patrol.gps

internal data class SuperGpsStreamOptions(
    val updateIntervalMs: Long = DEFAULT_UPDATE_INTERVAL_MS,
    val minUpdateIntervalMs: Long = DEFAULT_MIN_UPDATE_INTERVAL_MS,
    val minUpdateDistanceMeters: Float = DEFAULT_MIN_UPDATE_DISTANCE_METERS,
    val enableBarometer: Boolean = DEFAULT_ENABLE_BAROMETER,
) {
    companion object {
        const val DEFAULT_UPDATE_INTERVAL_MS = 700L
        const val DEFAULT_MIN_UPDATE_INTERVAL_MS = 500L
        const val DEFAULT_MIN_UPDATE_DISTANCE_METERS = 0f
        const val DEFAULT_ENABLE_BAROMETER = false

        fun fromArguments(arguments: Any?): SuperGpsStreamOptions {
            if (arguments !is Map<*, *>) return SuperGpsStreamOptions()

            return SuperGpsStreamOptions(
                updateIntervalMs = readLong(
                    arguments,
                    "updateIntervalMs",
                    DEFAULT_UPDATE_INTERVAL_MS,
                ),
                minUpdateIntervalMs = readLong(
                    arguments,
                    "minUpdateIntervalMs",
                    DEFAULT_MIN_UPDATE_INTERVAL_MS,
                ),
                minUpdateDistanceMeters = readFloat(
                    arguments,
                    "minUpdateDistanceMeters",
                    DEFAULT_MIN_UPDATE_DISTANCE_METERS,
                ),
                enableBarometer = readBool(
                    arguments,
                    "enableBarometer",
                    DEFAULT_ENABLE_BAROMETER,
                ),
            )
        }

        private fun readBool(map: Map<*, *>, key: String, default: Boolean): Boolean {
            val value = map[key] ?: return default
            return when (value) {
                is Boolean -> value
                else -> default
            }
        }

        private fun readLong(map: Map<*, *>, key: String, default: Long): Long {
            val value = map[key] ?: return default
            return when (value) {
                is Int -> value.toLong()
                is Long -> value
                is Double -> value.toLong()
                is Float -> value.toLong()
                else -> default
            }
        }

        private fun readFloat(map: Map<*, *>, key: String, default: Float): Float {
            val value = map[key] ?: return default
            return when (value) {
                is Int -> value.toFloat()
                is Long -> value.toFloat()
                is Double -> value.toFloat()
                is Float -> value
                else -> default
            }
        }
    }
}
